### PR TITLE
9658 api

### PR DESCRIPTION
**Pull request by withanage**

### CODE REVIEW

File: api/v1/invitations/InvitationController.php [hunk 1]
```diff
@@ -243,4 +260,29 @@ public function decline(Request $illuminateRequest): JsonResponse
     {
         return $this->selectedHandler->decline($illuminateRequest);
     }
+
+    public function getMany(Request $illuminateRequest): JsonResponse
+    {
+        $context = $illuminateRequest->attributes->get('context'); /** @var \PKP\context\Context $context */
```
Comment (withanage) [inline api/v1/invitations/InvitationController.php:266]: @ipula  I do not think, we need these comments.

File: classes/invitation/sections/Email.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+use PKP\mail\Mailable;
+use PKP\user\User;
+use stdClass;
+
+class Email extends Section
+{
+    public bool $anonymousRecipients = false;
+    public array $locales;
+    public Mailable $mailable;
+    /** @var array<User> */
```
Comment (withanage) [inline classes/invitation/sections/Email.php:19]: I  wpuld get rid of this comments

File: classes/invitation/sections/Form.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        $config->primaryLocale = $this->form->getConfig()['primaryLocale'];
+        $config->visibleLocales = $this->form->getConfig()['visibleLocales'];
+        $config->supportedFormLocales = $this->form->getConfig()['supportedFormLocales'];
+        $config->errors = $this->form->getConfig()['errors'];
+        ;
+
+        //        dd($this->form->getConfig()['id']);
+        // Decision forms shouldn't have submit buttons
+        // because the step-by-step decision wizard includes
+        // next/previous buttons
```
Comment (withanage) [inline classes/invitation/sections/Form.php:44]: remove this comments

File: classes/invitation/steps/Step.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    public string $name;
+    public string $description;
+    public string $stepLabel;
+    public string $nextButtonLabel;
+    public array $sections = [];
+
+    /**
+     * @param string $id A unique id for this step
+     * @param string $name The name of this step. Shown to the user.
+     * @param string $description A description of this step. Shown to the user.
```
Comment (withanage) [inline classes/invitation/steps/Step.php:20]: @param string $description A description of this step , shown to the user.

File: classes/middleware/ValidateCsrfToken.php [hunk 5]
```diff
@@ -37,11 +37,11 @@ public function handle(Request $request, Closure $next)
...
-        if (!$this->isCsrfValid($request)) {
-            return response()->json([
-                'error' => __('form.csrfInvalid')
-            ], Response::HTTP_FORBIDDEN);
-        }
+        //        if (!$this->isCsrfValid($request)) {
+        //            return response()->json([
+        //                'error' => __('form.csrfInvalid')
+        //            ], Response::HTTP_FORBIDDEN);
+        //        }
```
Comment (withanage) [inline classes/middleware/ValidateCsrfToken.php:44]: if unused,  I would remove this

File: templates/invitation/acceptInvitation.tpl [hunk 6]
```diff
@@ -0,0 +1,23 @@
+<?php
+{**
+ * templates/management/invitation.tpl
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2003-2021 John Willinsky
```
Comment (withanage) [inline templates/invitation/acceptInvitation.tpl:6]: copyright 2024

File: templates/invitation/userInvitation.tpl [hunk 7]
```diff
@@ -0,0 +1,24 @@
+?php
+{**
+ * templates/management/invitation.tpl
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2003-2021 John Willinsky
```
Comment (withanage) [inline templates/invitation/userInvitation.tpl:6]: copyright 2024

File: api/v1/invitations/InvitationController.php (hunk 1 above)
Comment (ipula) [reply api/v1/invitations/InvitationController.php:266]: this is from Dimithri's commit

File: classes/invitation/core/Invitation.php [hunk 8]
```diff
@@ -166,7 +187,7 @@ protected function checkForKey()
 
     public function setExpiryDate(Carbon $expiryDate)
     {
-        if ($this->getStatus() !== InvitationStatus::INITIALIZED) {
+        if ($this->getStatus() !== InvitationStatus::INITIALIZED) {+
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/core/Invitation.php:190]: image

[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/core/Invitation.php:190]: also from Dimithir's commit

File: classes/invitation/invitations/UserRoleAssignmentInvite.php [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+    public const INVITATION_TYPE = 'userRoleAssignment';
+
+    protected array $notAccessibleAfterInvite = [
+        'userGroupsToAdd',
+        'userGroupsToRemove',
+    ];
+
+    protected array $notAccessibleBeforeInvite = [
+        // 'orcid',
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/UserRoleAssignmentInvite.php:52]: also from Dimithir's commit

File: classes/invitation/invitations/UserRoleAssignmentInvite.php [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+            ->when(isset($this->invitationModel->userId), function ($query) {
+                return $query->byUserId($this->invitationModel->userId);
+            })
+            ->when(!isset($this->invitationModel->userId) && $this->invitationModel->email, function ($query) {
+                return $query->byEmail($this->invitationModel->email);
+            })
+            ->byContextId($this->invitationModel->contextId)
+            ->delete();
+
+        // foreach($pendingInvitations as $pendingInvitation) {
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/UserRoleAssignmentInvite.php:151]: also from Dimithir's commit

File: classes/invitation/invitations/UserRoleAssignmentInvite.php [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                    $this->removeUserGroup($userGroupArray, $group['userGroup']);
+                    $this->addUserGroup($userGroupArray, $group);
+                }
+            }
+        }
+    }
+
+    private function addUserGroup(array &$userGroupArray, array $userGroup)
+    {
+        // $userGroup['userGroupObject'] = Repo::userGroup()->get($userGroup['userGroup']);
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/UserRoleAssignmentInvite.php:248]: also from Dimithir's commit

File: classes/invitation/invitations/UserRoleAssignmentInvite.php [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    // }
+
+    /**
+     * @inheritDoc
+     */
+    public function validate(): bool
+    {
+        // Custom rules
+        if (isset($this->userGroupsToAdd)) {
+            if (empty($this->userGroupsToAdd)) {
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/UserRoleAssignmentInvite.php:289]: also from Dimithir's commit

File: classes/invitation/invitations/handlers/UserRoleAssignmentInviteRedirectController.php [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        //     null,
+        //     'user',
+        //     'profile',
+        //     [
+        //         'contact'
+        //     ]
+        // );
+
+        // $request->redirectUrl($url);
+    }
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/handlers/UserRoleAssignmentInviteRedirectController.php:79]: also from Dimithir's commit

File: classes/invitation/invitations/handlers/UserRoleAssignmentInviteRedirectController.php [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    //         $collector->filterByContextIds([$request->getContext()->getId()]);
+    //     } else {
+    //         $collector->withSiteAnnouncements(Collector::SITE_ONLY);
+    //     }
+
+    //     $announcements = $collector->getMany();
+
+    //     $templateMgr->assign('announcements', $announcements->toArray());
+    //     $templateMgr->display('frontend/pages/announcements.tpl');
+    // }
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/handlers/UserRoleAssignmentInviteRedirectController.php:112]: also from Dimithir's commit

File: classes/invitation/invitations/handlers/api/UserRoleAssignmentCreateController.php [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        $userGroupsToAdd = $payload['userGroupsToAdd'];
+        $this->invitation->updateUserGroupArray($this->invitation->userGroupsToAdd, $userGroupsToAdd);
+
+        $userGroupsToRemove = $payload['userGroupsToRemove'];
+        $this->invitation->updateUserGroupArray($this->invitation->userGroupsToRemove, $userGroupsToRemove);
+
+        $this->invitation->updatePayload();
+
+        if (!$this->invitation->isValid()) {
+            // This can be generalised inside the HasValidation trait
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/handlers/api/UserRoleAssignmentCreateController.php:83]: also from Dimithir's commit

File: classes/invitation/invitations/handlers/api/UserRoleAssignmentReceiveController.php [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        $reqInput = $illuminateRequest->all();
+        $payload = $reqInput['invitationData'];
+
+        $this->invitation->fillFromArgs($payload);
+
+        $this->invitation->updatePayload();
+
+        if (!$this->invitation->isValid()) {
+            // This can be generalised inside the HasValidation trait
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/handlers/api/UserRoleAssignmentReceiveController.php:104]: also from Dimithir's commit

File: classes/invitation/sections/Email.php (hunk 2 above)
Comment (ipula) [reply classes/invitation/sections/Email.php:19]: done

File: classes/invitation/invitations/UserRoleAssignmentInvite.php [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    {
+        $userGroupArray = array_filter($userGroupArray, function ($group) use ($userGroupId) {
+            return $group['userGroup'] != $userGroupId;
+        });
+
+        // Re-index the array to maintain a consistent array structure
+        $userGroupArray = array_values($userGroupArray);
+    }
+
+    // public function toJsonArray(): array
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/UserRoleAssignmentInvite.php:262]: also from Dimithir's commit

File: classes/invitation/invitations/handlers/api/UserRoleAssignmentCreateController.php [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+            ];
+
+            return response()->json(
+                $response,
+                Response::HTTP_UNPROCESSABLE_ENTITY
+            );
+        }
+
+        // Here we should consider returning a certain json taken from the custom invitation
+        // in order to be able to fully control the response
```
[the remark this answers is not in the document]
Comment (ipula) [reply classes/invitation/invitations/handlers/api/UserRoleAssignmentCreateController.php:97]: also from Dimithir's commit

File: classes/invitation/sections/Form.php (hunk 3 above)
Comment (ipula) [reply classes/invitation/sections/Form.php:44]: done

File: classes/invitation/steps/Step.php (hunk 4 above)
Comment (ipula) [reply classes/invitation/steps/Step.php:20]: done

File: classes/middleware/ValidateCsrfToken.php (hunk 5 above)
Comment (ipula) [reply classes/middleware/ValidateCsrfToken.php:44]: done

File: templates/invitation/acceptInvitation.tpl (hunk 6 above)
Comment (ipula) [reply templates/invitation/acceptInvitation.tpl:6]: done

**Pull request closed without merging**